pkg/git/libgit2/transport.go

```diff
@@ -126,20 +128,40 @@ func (s *PublicKeyAuth) Method(secret corev1.Secret) (*git.Auth, error) {
 		user = git.DefaultPublicKeyAuthUser
 	}
 
-	credCallback := func(url string, username_from_url string, allowed_types git2go.CredType) (*git2go.Cred, error) {
+	credCallback := func(url string, usernameFromURL string, allowedTypes git2go.CredType) (*git2go.Cred, error) {
 		cred, err := git2go.NewCredSshKeyFromMemory(user, "", string(identity), "")
 		if err != nil {
 			return nil, err
 		}
 		return cred, nil
 	}
 	certCallback := func(cert *git2go.Certificate, valid bool, hostname string) git2go.ErrorCode {
+		// First, attempt to split the configured host and port to validate
+		// the port-less hostname given to the callback.
+		host, _, err := net.SplitHostPort(s.host)
+		if err != nil {
+			// SplitHostPort returns an error if the host is missing
+			// a port, assume the host has no port.
+			host = s.host
+		}
+
+		// Check if the configured host matches the hostname given to
+		// the callback.
+		if host != hostname {
+			return git2go.ErrUser
+		}
+
+		// We are now certain that the configured host and the hostname
+		// given to the callback match. Use the configured host (that
+		// includes the port), and normalize it so we can check if there
+		// is an entry for the hostname _and_ port.
+		host = knownhosts.Normalize(s.host)
 		for _, k := range kk {
-			if k.matches(hostname, cert.Hostkey.HashSHA1[:]) {
+			if k.matches(s.host, cert.Hostkey.HashSHA1[:]) {
```

Comment:
This should use the normalised `host` for comparisons rather than the original `s.host`

Reply:
This PR is cursed :S